### PR TITLE
Present OpenAPI `additionalProperties` field

### PR DESCRIPTION
**Pull request by DMRobertson**
When looking at the "Authentication Data" type used in e.g. the account
deactivation endpoint

    https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3accountdeactivate

it was not clear to me that clients and servers should expect additional
fields in this structure (the semantics of which are implied by the
`type` of the authentication, or that of the previously established
`session`.)

In the OpenAPI spec, We occasionally mark some ojects as allowing
arbitrary additional properties (`additionalProperties: true`, or
`additionalProperties: { description: "..."}`). In other places we are
more strict and provide a schema that additional properties must satisfy.
In this PR we aim to make the first kind of additional
properties (non-strict) more visible in the rendered spec.

See https://pr1162--matrix-spec-previews.netlify.app/client-server-api/#post_matrixclientv3accountdeactivate

image

Preview: https://pr1162--matrix-spec-previews.netlify.app

### CODE REVIEW

File: layouts/partials/openapi/render-object-table.html [hunk 1]
```diff
@@ -93,6 +94,29 @@
...
+May have additional properties: {{ $additionalProperties.description }}
+            {{ end }}
+        {{/* TODO: should we handle the case where additional properties are permitted,
+                    but must follow an explicit schema? */}}
+        {{ else if $additionalProperties -}}
+May have additional properties.
+        {{ else if (not $additionalProperties) -}}
+        {{/* At present we don't have any schema with additionalProperties: false.
+        But if we ever do, let's support it. */}}
+May **not** have additional properties.
```
Comment (turt2live) [inline layouts/partials/openapi/render-object-table.html:115]: Everything is extensible in Matrix unless otherwise noted, so this is a bit misleading.

**Review by turt2live — COMMENTED** [review]
generally this seems fine, though there seems to be a bit of confusion around what clients are able to send (which is anything it feels like - servers don't have to listen though)

File: layouts/partials/openapi/render-object-table.html (hunk 1 above)
Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:115]: I don't follow---what do you find misleading?

Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:115]: Okay. Then perhaps this can be changed so that 

- we don't do anything if `additionalProperties: true`
- we raise a build-time error if `additionalProperties: false`
- Rather than "May have additional properties:" in the generated HTML, we write "Notes on additional properties:".

How does that sound?

Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:115]: See https://swagger.io/specification/#schema-object

> The following properties are taken from the JSON Schema definition but their definitions were adjusted to the OpenAPI Specification.
> additionalProperties - Value can be boolean or object. Inline or referenced schema MUST be of a [Schema Object](https://swagger.io/specification/#schema-object) and not a standard JSON Schema. Consistent with JSON Schema, additionalProperties defaults to true.

Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:115]: And for completeness: the original JSON schema additionalProperties are defined [here](https://json-schema.org/draft/2020-12/json-schema-core.html#additionalProperties).

Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:112]: 3398cf0f is an attempt at this.

**Comment by DMRobertson** [comment]
Oh. Looks like I have a merge to resolve.

File: layouts/partials/openapi/render-object-table.html [hunk 2]
```diff
@@ -18,6 +18,7 @@
 {{ $title := .title }}
 {{ $properties := .properties}}
 {{ $required := .required}}
+{{ $additionalProperties := .additionalProperties }}
```
Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:21]: please could you add something about this to the comment at the top of the file

File: layouts/partials/openapi/render-object-table.html [hunk 3]
```diff
@@ -97,6 +98,26 @@
 
     {{ end }}
 
+ {{/* Additional properties should either be an OpenAPI schema object, or a
```
Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:101]:
```suggestion
 {{/* additionalProperties should either be an OpenAPI schema object, or a
```

File: layouts/partials/openapi/render-object-table.html [hunk 4]
```diff
@@ -97,6 +98,26 @@
...
+       boolean. If it's the former, look for a `description` and render it as a note.
+       If it's `true`, do nothing. If it's `false`, raise a build-time error:
+       "Everything is extensible in Matrix unless otherwise noted". */}}
+   {{ if (ne $additionalProperties nil )}}
+  <tr class="row-additional-properties">
+   <td colspan="3">
```
Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:107]: it seems odd that we include the table row and empty cell if `additionalProperties` is true?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: layouts/partials/openapi/render-object-table.html (hunk 2 above)
Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:30]: Sure, see 6487e84c

File: layouts/partials/openapi/render-object-table.html (hunk 4 above)
Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:116]: Yes. A previous version of this PR used to have the cell say "May have additional properties" in that situation. Looks like I didn't clean up the HTML for when I removed that case.

File: layouts/partials/openapi/render-object-table.html [hunk 5]
```diff
@@ -97,6 +107,27 @@
...
     {{ end }}
 
+ {{/* If additionalProperties is true, do nothing.
+      If additionalProperties is nil, treat is as the default (true) and do nothing.
+      If additionalProperties is false, raise a build-time error:
+      "Everything is extensible in Matrix unless otherwise noted".
+      Otherwise, it should be an OpenAPI schema object. Include the description
+      of that schema object, if it exists. */}}
+ {{ if (and (ne $additionalProperties nil) (ne $additionalProperties true)) }}
+  {{ if (and (reflect.IsMap $additionalProperties)) -}}
```
Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:117]: redundant `and` here.

Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:117]: can we restructure these `if` statements to reduce the nesting?

```
 {{ if (or (eq $additionalProperties nil) (eq $additionalProperties true)) }}
     {{/* do nothing */}}
  {{ else if (reflect.IsMap $additionalProperties) -}}
     ...
  {{ else if (not $additionalProperties) -}}
    ... 
  {{ end }}
```

or something.

File: layouts/partials/openapi/render-object-table.html [hunk 6]
```diff
@@ -97,6 +107,27 @@
...
+  <tr class="row-additional-properties">
+   <td colspan="3">
+    <em>Notes on additional properties:</em> {{ $additionalProperties.description }}
+   </td>
+  </tr>
+   {{ end }}
+  {{/* TODO: should we handle the case where additional properties are
+             permitted, but must follow an explicit schema? */}}
+  {{ else if (not $additionalProperties) -}}
+   {{ errorf "Unexpected additionalProperties=false for %s" $title }}
```
Comment (richvdh) [inline layouts/partials/openapi/render-object-table.html:128]: it's unclear to me that we should be raising an error in this case. Isn't it legitimate to specify that a given object may *not* have additional properties, even if that is currently unused?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: layouts/partials/openapi/render-object-table.html (hunk 6 above)
Comment (DMRobertson) [reply layouts/partials/openapi/render-object-table.html:128]: :shrug: There's some context here: https://github.com/matrix-org/matrix-spec/pull/1162#discussion_r914153682 ; I don't have any strong opinions here.

File: layouts/partials/openapi/render-object-table.html (hunk 1 above)
Comment (richvdh) [reply layouts/partials/openapi/render-object-table.html]: @turt2live, @DMRobertson: I'm not sure I'm following the discussion here. Can I just check that everyone is clear what we're discussing?

The question is around the correct behaviour if we encounter an object schema where `additionalProperties` is explictly set to `false` (as opposed to leaving it unset, which is equivalent to setting it to `true`).

[Earlier drafts](https://datatracker.ietf.org/doc/html/draft-wright-json-schema-validation-00#section-5.18) of JSON Schema give a clear definition of what that means:

> If "additionalProperties" is false, validation succeeds only if the instance is an object and all properties on the instance were covered by "properties" and/or "patternProperties".

In other words: this is explicitly about trying to define an object which **may not be extended**.

I don't entirely *object* to the decision to emit an error message, if we think it's somehow un-matrixy to try to define an object that isn't allowed any additional properties.  I'd just like to be sure that we made that decision with good information.

Comment (KitsuneRal) [reply layouts/partials/openapi/render-object-table.html]: We actually have that case (exactly one, unless I missed some weirdly formatted), and it's in [`m.receipt` events](https://github.com/matrix-org/matrix-spec/blob/a45138c3b4a722ef4f4eb98788dc25c94a3dce61/data/event-schemas/schema/m.receipt.yaml#L58): . `additionalProperties: false` is used next to `patternProperties` to forbid anything that does not follow the pattern. Which is quite valid and smart, in my opinion. There's a big question however on what the homeserver is supposed to do with this boundary. I would treat it as a "MUST ignore" of sorts, whereas omitting `additionalProperties` entirely would correspond to "MAY ignore". Does it make sense?

**Comment by richvdh** [comment]
@DMRobertson are you still interested in pursuing this? Anything we can do to unblock you?

**Comment by DMRobertson** [comment]
It'd be nice to do this at some point, but I don't think I'm going to find the time. Let's close this off for now.
